ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java

```diff
@@ -1737,6 +1747,14 @@ public RelNode apply(RelOptCluster cluster, RelOptSchema relOptSchema, SchemaPlu
       perfLogger.perfLogBegin(this.getClass().getName(), PerfLogger.POSTJOIN_ORDERING);
       calcitePlan = applyPostJoinOrderingTransform(calcitePlan, mdProvider.getMetadataProvider(), executorProvider);
       perfLogger.perfLogEnd(this.getClass().getName(), PerfLogger.POSTJOIN_ORDERING);
+      // Perform the CTE rewriting near the end of CBO transformations to avoid interference of the new HiveTableSpool 
+      // operator with other rules (especially those related to constant folding and branch pruning).
```

Comment:
It would be useful to have a simple q test with branch pruning on both sides of a UNION ALL (both sides have CTE with FALSE predicate).. in that case the  branches should be pruned first and the CTE part should be skipped altogether.

Reply:
In fact it is more problematic when on side of the union is pruned and the other remains. The actual risk is to have one branch with a CTE scan with no corresponding spool. I added corresponding tests in  https://github.com/apache/hive/pull/5249/commits/c62adf3fe26035a3f2d8857292de595d861be927